FiltersEngine.cpp

```diff
@@ -0,0 +1,201 @@
+{
+    if (pfs._depth != -1 && pfs._depth < height) {
+        return false;
+    }
+
+    if (pfs._match_mask & PFS_MATCH_UID) {
+        if (pfs._uid != process->_uid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_GID) {
+        if (pfs._gid != process->_gid) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_EQUALS) {
+        if (pfs._exeMatchValue != process->_exe) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_STARTSWITH) {
+        if (process->_exe.substr(0, pfs._exeMatchValue.length()) != pfs._exeMatchValue) {
+            return false;
+        }
+    }
+
+    if (pfs._match_mask & PFS_MATCH_EXE_CONTAINS) {
+        if (process->_exe.find(pfs._exeMatchValue) == std::string::npos) {
+            return false;
+        }
+    }
+    if (pfs._match_mask & PFS_MATCH_EXE_REGEX) {
+        if (!std::regex_search(process->_exe, pfs._exeRegex)) {
+            return false;
+        }
+    }
+
+    for (auto cf : pfs._cmdlineFilters) {
+        if (cf._matchType == MatchEquals) {
+            if (cf._matchValue != process->_cmdline) {
+                return false;
+            }
+        } else if (cf._matchType == MatchStartsWith) {
+            if (process->_cmdline.substr(0, cf._matchValue.length()) != cf._matchValue) {
+                return false;
+            }
+        } else if (cf._matchType == MatchContains) {
+            if (process->_cmdline.find(cf._matchValue) == std::string::npos) {
+                return false;
+            }
+        } else if (cf._matchType == MatchRegex) {
+            if (!std::regex_search(process->_cmdline, cf._matchRegex)) {
+                return false;
+            }
+        }
+    }
+
+    return true;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::GetFlags(std::shared_ptr<ProcessTreeItem> process, unsigned int height)
+{
+    std::bitset<FILTER_BITSET_SIZE> flags;
+
+    for (auto element : _filtersBitPosition) {
+        ProcFilterSpec pfs = element.first;
+        if (ProcessMatchFilter(process, pfs, height)) {
+            flags[element.second.bitPosition] = 1;
+        }
+    }
+
+    return flags;
+}
+
+std::bitset<FILTER_BITSET_SIZE> FiltersEngine::GetCommonFlagMask()
+{
+    std::bitset<FILTER_BITSET_SIZE> flags;
+    unsigned int numberOfOutputs = _outputs.size();
+
+    for (auto element : _filtersBitPosition) {
+        if (element.second.outputs.size() == numberOfOutputs) {
+            flags[element.second.bitPosition] = 1;
+        }
+    }
+
+    return flags;
+}
+
+bool FiltersEngine::syscallIsFiltered(std::string& syscall, std::vector<std::string>& syscalls)
+{
+    for (auto s : syscalls) {
+        if (s == "*") {
+            return true;
+        } else if (s == syscall) {
+            return true;
+        } else if (s == "!" + syscall) {
```

Comment:
Nice I was thinking of using the ! for Sysmon on Windows too. For AUOMS is this just for syscall filters? The Process Match filters didn't seem to support it?

Reply:
ProcessMatchFilters just works out which processes to check the syscalls on. This information gets stored as flags on the process in the ProcessTree. Later when an event happens, the process is matched in the tree, the flags are retrieved and the corresponding syscall filters for the set flags are checked with syscallIsFiltered.
It's basically a half-and-half solution where up front we annotate processes on creation, and only check the syscalls on events.

Reply:
Sorry I see what you're asking. Yes, this is just for syscalls. It's a way to cause certain syscalls to always send events.